File: Python/ASP Advanced/Graphs/Topological Sort/Leetcode-1203.py
from collections import defaultdict

class Solution:
    def sortItems(self, n, m, group, beforeItems):
        for i in range(n):
            if group[i] == -1:
                group[i] = m
                m += 1

        item_adj = defaultdict(list)
        group_adj = defaultdict(list)
        for i in range(n):
            for par in beforeItems[i]:
                item_adj[par].append(i)
                if group[i] != group[par]:
                    group_adj[group[par]].append(group[i])

        print(item_adj)
        print(group_adj)

        itm = self.topo_sort(item_adj, n)
        if not itm: return []
        grp = self.topo_sort(group_adj, m)
        if not grp: return []

        print(itm)
        print(grp)

        grouping = defaultdict(list)
        for i in itm:
            grouping[group[i]].append(i)

        print(grouping)

        res = []
        for g in grp:
            res.extend(grouping[g])

        return res

    def topo_sort(self, adj, N):
        visited = [0] * N
        topo = []

        def dfs(node):
            if visited[node] == 1:
                return True
            if visited[node] == 2:
                return False
            visited[node] = 1
            for neighbor in adj[node]:
                if dfs(neighbor):
                    return True
            topo.append(node)
            visited[node] = 2
            return False

        for i in range(N):
            if visited[i] == 0:
                if dfs(i):
                    return []

        return topo[::-1]

sol = Solution()
n = 8
m = 2
group = [-1,-1,1,0,0,1,0,-1]
beforeItems = [[],[6],[5],[6],[3,6],[],[],[]]
print(sol.sortItems(n, m, group, beforeItems))

File: C++/ASP Advanced/Graphs/Topological Sort/Leetcode-1203.cpp
class Solution {
private:
    unordered_map<int, vector<int>> mapNodeGroup;
    unordered_set<int> visited;
    vector<int> topoSortGroups;
    vector<int> finalTopo;
    unordered_map<int, vector<int>> adjGroup;

    // topological sort for groups
    unordered_set<int> path;
    bool topoSortGroup(int groupNode){
        if (path.count(groupNode)) return false;
        if (visited.count(groupNode))return true;

        path.insert(groupNode);
        for (auto nei: adjGroup[groupNode]){
            if (!topoSortGroup(nei)){
                return false;
            }
        }
        path.erase(groupNode);
        visited.insert(groupNode);
        topoSortGroups.push_back(groupNode);
        return true;
    }

    // topological sort for node in the same group
    bool topo(int node, int myGroup, vector<int>& group, vector<vector<int>>& beforeItems){
        if (path.count(node)) return false;
        if (visited.count(node))return true;

        if (myGroup == group[node]) visited.insert(node);
        path.insert(node);
        for (auto nei: beforeItems[node]){
            if (!topo(nei, myGroup, group, beforeItems)){
                return false;
            }
        }
        path.erase(node);
        if (myGroup == group[node]) finalTopo.push_back(node);

        return true;
    }

public:
    vector<int> sortItems(int n, int m, vector<int>& group, vector<vector<int>>& beforeItems) {
       
        for (int g = 0; g < n; g++){
            if (group[g] == -1){
                group[g] = m++;
            }
            mapNodeGroup[group[g]].push_back(g);
        }

        for (int g = 0; g < n; g++){
            for (auto b: beforeItems[g]){
                if (group[g] != group[b])
                    adjGroup[group[g]].push_back(group[b]);
            }
        }

        for (int g = 0; g < m; g++){
            if (!topoSortGroup(g)){
                return vector<int>();
            }
        }
        path.clear();
        visited.clear();
        for (int i = 0; i < topoSortGroups.size(); i++){
            int g = topoSortGroups[i];
            cout << "g: " << g << " elems: ";
            for (auto node: mapNodeGroup[g]){
                cout << node << " ";
                if (!topo(node, g, group, beforeItems)){
                    return vector<int>();
                }
            }
            cout << endl;
        }

        return finalTopo;
    }
};